stl/inc/xstring

```diff
@@ -2195,6 +2195,12 @@ public:
 template <class _Ty>
 constexpr size_t _Size_after_ebco_v = is_empty_v<_Ty> ? 0 : sizeof(_Ty); // get _Ty's size after being EBCO'd
 
+struct _String_constructor_concat_tag {};

```

Comment:
Idea: I observe that you've made this implicitly default constructible, while the `basic_string` ctor is `explicit`. What if you gave this tag an `explicit` default constructor (as some tags in the Standard do), forbidding `{}` from being used to construct the tag, and then made the `basic_string` ctor implicit? That would allow you to say `return { _String_constructor_concat_tag{}, ARGS }` below, while not seriously affecting safety in any way. Feel free to disregard though.

Reply:
I made the `basic_string` ctor `explicit` because I want to be as paranoid as possible about conversion sequences affecting `basic_string`'s public interface, and I dislike giving these tags ctors because our ABI won't pass such a tag in a register. However, it looks like if we make an inline constexpr variable in release mode that gets eaten so that's OK: https://gcc.godbolt.org/z/jAU-cr 

Hmmm and for some reason it looks like in debug mode that's actually an improvement: I find that surprising. https://gcc.godbolt.org/z/0IHz4F Maybe we should start doing this for all tags (including one_then_variadic_args and friends)?

Reply:
(That is to say, my initial reaction was "I don't want to do that because codegen", then I actually checked the codegen, and now my reaction is "I want to do that because codegen" :)

Reply:
Can you add that suggestion (`inline constexpr`) to #468 and explain your codegen discovery? We didn't have `inline constexpr` back when we implemented the old tags. (Also, we should check whether C++14 without `inline` variables tolerates this as well.)

Reply:
Will do

Reply:
> However, it looks like if we make an inline constexpr variable in release mode that gets eaten so that's OK

You need to uncheck the ".LX0" and ".text" filters in CE to see the definition of the variable; it's not emitted in the non-`USE_INLINE` cases since it isn't ODR-used, but is emitted for `/DUSE_INLINE`. The compiler apparently doesn't realize that the object representation is only padding bits that it need not copy when making a copy of the variable; it copies a byte of uninitialized stack space for non-`USE_INLINE`, and copies the representation `tagInstance` for `USE_INLINE`.

Reply:
Ah, Casey is right. In that case release codegen for creating the temporary (what we already do) is best.